Include token indexes in query tuning section (#1459)
neo4j/neo4j-documentation: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc

== Index types and predicate compatibility

> Generally, an index solve some combination of a label predicate and a property predicate at the same time.

Reviewer: ```suggestion
Generally, an index solves some combination of a label predicate and property predicates at the same time.
```

> There are different types of indexes available in Neo4j but they are not all compatible with the same property predicates.

Reviewer: Opinionated:
```suggestion
There are different types of indexes available in Neo4j and these are compatible with different property predicates.
```

Author: Agreed

Indexes are most often used for `MATCH` and `OPTIONAL MATCH` clauses that combine a label predicate with a property predicate.

> Therefore, it is important to know what kind of predicates that can be solved by the different indexes.

Reviewer: ```suggestion
Therefore, it is important to know what kind of predicates can be solved by the different indexes.
```

> An important exception to this is `TOKEN LOOKUP` indexes, which solve only label predicates.

Reviewer: ```suggestion
An important exception to this are `TOKEN LOOKUP` indexes, which solve only label predicates.
```

Reviewer 2: I think this is sentence is unnecessary as the next one is exactly the same.

=== TOKEN LOOKUP indexes

`TOKEN LOOKUP` indexes are present by default and solve only label and relationship type predicates:

[options="header"]
|===
|Label and relationship type predicates
|`MATCH (n:Label)`
|`MATCH (n) WHERE n:Label`
|`MATCH ()-[r:REL]\->()`
|`MATCH ()-[r]->() WHERE r:REL`
|===

> The lookup indexes are the most important indexes in the database and greatly improves the performance of Cypher queries, as well as the population of other indexes.

Reviewer: ```suggestion
The lookup indexes are the most important indexes in the database and greatly improve the performance of Cypher queries, as well as the population of other indexes.
```

> Dropping these indexes leads to severe performance degradation for most workloads.

Reviewer: ```suggestion
Dropping these indexes can lead to severe performance degradation for most workloads.
```
Should we also write, when one might do that? E.g. to increase write throughput.

* `TEXT` indexes are preferred over `BTREE` indexes for `CONTAINS` and `ENDS WITH`.
* `BTREE` indexes are preferred over `TEXT` indexes in all other cases.

> `TOKEN LOOKUP` indexes are not defined in this order since they never solve the same set of predicates as the other indexes.

Reviewer: Especially this should be synced with the changes done for documenting the new index types.

* `POINT` indexes are used over `RANGE` and `TEXT` indexes for distance and within a bounding box.
* `RANGE` indexes are preferred over `TEXT` and `POINT` indexes in all other cases.

`TOKEN LOOKUP` indexes are not defined in this order since they never solve the same set of predicates as the other indexes.

include::ql/administration/indexes/node-label-lookup-index-example.asciidoc[leveloffset=2]

include::ql/administration/indexes/relationship-type-lookup-index-example.asciidoc[leveloffset=2]

> include::ql/administration/indexes/relationship-range-index-example.asciidoc[leveloffset=2]

Reviewer 3: This is included twice...

`include::ql/administration/indexes/relationship-range-index-example.asciidoc[leveloffset=2]`
